source/agora/consensus/state/ValidatorSet.d

```diff
@@ -386,38 +352,30 @@ public class ValidatorSet
 
     /***************************************************************************
 
-        Clear up expired validators whose cycle for a validator ends
+        Clear up expired validators from the cycle before the previous cycle
 
         The validator set clears up expired validators from the set based on
-        the block height. Validators are deleted if their enrolled height is
-        less than or equal to the value of the passed block height minus the
-        validator cycle.
+        the block height. We need to keep the previous enrollment record so that
+        we can still verify signatures received after the end of a cycle.
 
         Params:
-            block_height = current block height
+            height = current block height
 
     ***************************************************************************/
 
-    public void clearExpiredValidators (Height block_height) @safe nothrow
+    public void clearExpiredValidators (in Height height) @safe nothrow
     {
         // the smallest enrolled height would be 0 (genesis block),
         // so the passed block height should be at minimum the
         // size of the validator cycle
-        if (block_height < this.params.ValidatorCycle)
+        if (height < this.params.ValidatorCycle)
             return;
 
         try
         {
             () @trusted {
-                if (block_height > this.params.ValidatorCycle)
-                {
-                    this.db.execute("DELETE from validator_set " ~
-                    "WHERE (enrolled_height < ? AND active = ?) or (enrolled_height < ?)",
-                    EnrollmentStatus.Expired, block_height - this.params.ValidatorCycle,
-                    block_height - this.params.ValidatorCycle - 1);
-                }
-                this.db.execute("UPDATE validator_set SET active = ? WHERE enrolled_height <= ? AND active = ?",
-                    EnrollmentStatus.Expired, block_height - this.params.ValidatorCycle, EnrollmentStatus.Active);
+                this.db.execute("DELETE from validator_set WHERE enrolled_height <= ?",
+                    minEnrollmentHeight(Height(height - this.params.ValidatorCycle)));
```

Comment:
```suggestion
                    this.minEnrollmentHeight(Height(height - this.params.ValidatorCycle)));
```

As mentioned earlier, I don't see the point of deleting if we always use `enroll_height`. And if we delete, I don't see the point of keeping `enroll_height` as part of the PK.

Reply:
Refactored now so that it keeps the old enrollments but uses height to determine if enrolled.